### ISSUE TITLE
Common default return value of functions with generics

### ISSUE THREAD

**Issue by balazsbonis** [FakeItEasy/FakeItEasy#648]
I am writing unit tests for little tasks that are talking to the database. My test cases also include scenarios where a given object is not found in the database, and therefore I'd like to emulate that the queries return `null`.

Now, I managed to achieve this by using the following calls:

```c#
var _session = // database session
A.CallTo(_session).WithReturnType<Foo>().Returns(null);
A.CallTo(_session).WithReturnType<Bar>().Returns(null);
// ~10 more of these with just different types defined
```

however I was wondering if there is a way of setting this up as the default behaviour - so that all calls to `_session` with a generic parameter will return `null` by default?

I don't know if it is possible in the first place, but it would probably be a nice feature.

**Comment by blairconrad** [comment]
This issue came from StackOverflow question [How to set the return value of multiple generic functions with FakeItEasy?][question].

[question]: http://stackoverflow.com/questions/36547346/how-to-set-the-return-value-of-multiple-generic-functions-with-fakeiteasy

**Comment by blairconrad** [comment]
I couldn't figure out a way to do this. I think we'd need something like a `WithNonVoidReturnType` or `WhenReturnTypeMatches` addition to the `A.CallTo(fake)` use case.

**Comment by blairconrad** [comment]
@balazsbonis, thanks for the idea. Assuming the other owners are also in favour of the addition, are you interested in working on it? The best course would be to discuss an API here, and if/once it's agreed on, to proceed with an implementation.

**Comment by thomaslevesque** [comment]
I like `WithNonVoidReturnType`. It should be pretty easy to add:

    IAnyCallConfigurationWithReturnTypeSpecified<object> WithNonVoidReturnType();

Of course, with this it becomes possible to return a value of the wrong type, so it will fail when the method is called...

**Comment by blairconrad** [comment]
Oh, agreed. It's not typesafe. Just like many of the other more flexible/powerful methods.

**Comment by adamralph** [comment]
What about:

```C#
A.CallTo(_session).ReturnsDefaultValue();
```

I.e. we just switch off the returning of automatic fakes for that object. We could also make it work with `Where()` for more control:

```C#
A.CallTo(_session).Where(call => ...).ReturnsDefaultValue();
```

**Comment by blairconrad** [comment]
@adamralph, I don't think @thomaslevesque's suggestion precludes using `Where`, and it would allow (assuming an appropriate non-typed `Returns` and `ReturnsLazily`) the client to set up any return value. I think it's preferable to allow the client to pick the value, instead of just limiting ourselves to defaults.

**Comment by blairconrad** [comment]
To expand, the way I thought @thomaslevesque's suggestion would work would be

```c#
A.CallTo(fake).Where(…).WithNonVoidReturnType().Returns(null);
```

(or swap `Where` and `WithNonVoidReturnType`, as desired, like we can now with `Where` and `WithReturnType<>`.)

**Comment by adamralph** [comment]
What would be the use case for a non-default value? Would it not be covered by `WithReturnType()`?

**Comment by blairconrad** [comment]
Someone may have a pre-existing factory method for making their preferred values?

**Comment by adamralph** [comment]
After [discussion with @blairconrad](https://gitter.im/FakeItEasy/FakeItEasy?at=570bc7138b7b2f457635924b), I'm :+1: for `WithNonVoidReturnType()`.

**Comment by balazsbonis** [comment]
@blairconrad I have never seen FakeItEasy's code yet, so I don't know how slow would it take me to actually implement the changes. Therefore if you already have a nice design of what exactly needs to be changed, I say you go ahead. I really like the idea you came up with. :+1:

**Comment by adamralph** [comment]
@balazsbonis speed isn't really important. This is a non-breaking change and it can be added at any time. If you'd like to take the time to learn the code base and make the change, we'd be more than happy to wait for it.

**Comment by blairconrad** [comment]
@balazsbonis, @adamralph beat me to it. We're in no rush for the feature, and if it benefits you and you'd enjoy working on it, we'd love to have your submission. And we'd be happy to talk through any questions you'd have, or give any hints, if you're feeling stuck.

**Comment by blairconrad** [comment]
Oh, and just so it doesn't look like I'm taking credit, @thomaslevesque came up with the syntax.
English's "you" is confusing, as you (see what I did there?) can't always tell whether it's singular or plural.

**Comment by thomaslevesque** [comment]
> Oh, and just so it doesn't look like I'm taking credit, @thomaslevesque came up with the syntax.

Actually, @blairconrad came up with the name :wink: 

I looked at the code, it should be pretty easy to implement. Great feature to discover the code base @balazsbonis :smile:

**Comment by blairconrad** [comment]
@thomaslevesque

> Actually, @blairconrad came up with the name :wink:

Oh. I guess so. I need my "embarrassed" emoji!

**Comment by adamralph** [comment]
Marked as taken, since @balazsbonis has started work on this in #674.

**Comment by adamralph** [comment]
Still needs docs.

**Comment by blairconrad** [comment]
And I wouldn't say "no" to a diagnostic that notices when we're not using the result.
This could be a separate issue or PR if we want.

**Comment by adamralph** [comment]
Nice idea. I'd be happy with it being a separate issue. I guess the same applies to `WithReturnType<TMember>()`?

**Comment by blairconrad** [comment]
>  the same applies to `WithReturnType<TMember>()`?

[already covered](https://github.com/FakeItEasy/FakeItEasy/blob/e2562dc0e5d2cb918e748b8223636748adc485ee/src/FakeItEasy.Analyzer/UnusedReturnValueAnalyzer.cs#L79)

**Comment by adamralph** [comment]
Oh! That easy! Well, let's just add it then.

But! Should it be a separate `enhancement` issue also labelled `analyzer`?

**Comment by adamralph** [comment]
What shall we do about the analyzer? I don't think we need a separate issue, we can just say that this feature complete with docs, analyzer warning, etc.

**Comment by thomaslevesque** [comment]
> What shall we do about the analyzer? I don't think we need a separate issue, we can just say that this feature complete with docs, analyzer warning, etc.

Yes, I'll reopen and submit a PR with the analyzer changes.

**Comment by blairconrad** [comment]
> we can just say that this feature comes complete with docs, analyzer warning, etc.

That's my preference in all things. It's part of the reason that any time I get an opportunity, I say "let's move it into the repo" where "it" could be docs, release notes, or whatever.

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.1.0).

Thanks for your help, @balazsbonis. Look for your name in the release notes! :trophy:
